docs/spec/governance/software_upgrade.md

```diff
@@ -0,0 +1,133 @@
+## Software Upgrade
+
+Software is upgraded according to a 4-step process.
+
+1. Agree on software upgrade plan via governance (currently via a plaintext proposal).
+2. Once the upgrade plan is accepted, validators each update their software according to proposal from step 1.
+3. Each upgraded validator submits a transaction to the SoftwareUpgradeKeeper.
+4. Once a sufficient quorum of validators has completed step 3, new logic kicks
+   in in the form of a special parameter.
+
+### Step 1: Agree on the Upgrade Plan
+
+First, an English plaintext proposal should be submitted based on the following template:
+
+```
+# SOFTWARE UPGRADE PROPOSAL
+
+This is a proposal to upgrade the software.
+
+## Outline:
+
+* Current version: [XXX]
+* Current version commit hash: [XXX]
+* New version Github URL: https://github.com/...
+* New version: [XXX]
+* New version commit hash: [XXX]
+* Upgrade prepared by:
+ - <name/handle> on <date>; commit signature: https://github.com/...
+* Audited by:
+ - <name/handle> on <date>; report: https://github.com/...
+ - <name/handle> on <date>; report: https://github.com/...
+
+## Description:
+
+Describe the purpose of the upgrade and other details.
+
+- [ ] Link to any previous discussions
+- [ ] Link to any alternative proposals
+
+## Requirements:
+
+Describe any requirements
+
+- [ ] Any new hardware requirements?
+- [ ] Any new bandwidth requirements?
+- [ ] Any dependencies on other proposals?
+
+## Procedure:
+
+* Deadline: By 2 weeks from after the proposal has been accepted.
+
+Once this proposal has been accepted, validators should upgrade to the new
+software as described above by the deadline.  Once the new software is
```

Comment:
A question here, how can we ensure that the version of software that upgraded to is identical to the version proposed because the proposal is not structured.